pylint/extensions/_check_docs_utils.py

```diff
@@ -164,12 +162,20 @@ def possible_exc_types(node):
                     and isinstance(val, (astroid.Instance, nodes.ClassDef))
                     and utils.inherit_from_std_ex(val)
                 ):
-                    excs.append(val.name)
+                    excs.append(val)
 
-    try:
-        return {exc for exc in excs if not utils.node_ignores_exception(node, exc)}
-    except astroid.InferenceError:
-        return set()
+    final_exceptions = set()
+    for exc in excs:
+        try:
+            # can fail because it is trying to infer the handler of the exception
+            if utils.node_ignores_exception(node, exc.name):
+                continue
+        except astroid.InferenceError:
+            return set()
+        if not isinstance(exc, nodes.ClassDef):
```

Comment:
Since you force-pushed I can't respond to the previous comment, but:

This doesn't really follow the previous logic as this was not part of the original `try ... except`. We need this because of changes to `excs` and thus I think we should handle this corner case earlier.

Reply:
Sorry for the forced push. I'll check

Reply:
So, the code is not longer part of the try-except. What would be the benefit for handling this earlier?

Reply:
It makes the corner case more specific. You had to add this because of some currently unknown corner case. Documenting and handling such corner cases explicitly is important for future maintainability of the code. If we ever want to refactor this code we won't know what this is doing which would make that much more difficult.

Reply:
So what is your suggestion? Can you please point me to the line where you see that and in which form. Personally I think the place is fine, but of course I don't have the big picture of pylint in mind

Reply:
Somewhere in between L137-165 as that is where we're actually determining the possible exception types.
(The original) L169-172  is only used to check for `utils.node_ignores_exception` for each possible node.

I think it would be if we could first determine what the actual corner case is that this is catching and then see where we should "handle" it. Determining the corner case would also allow us to add a comment explaining what the code is trying to do.

Reply:
I added a comment why this is needed. In fact `pytest tests/test_functional.py -k test_functional[missing_raises_doc]` will fail if the line is not there

Reply:
I have investigated this myself and the following diff would allow us to remove this `if` statement:
We can then revert most of these lines to the previous version and only add `.name`.

```diff
diff --git a/pylint/extensions/_check_docs_utils.py b/pylint/extensions/_check_docs_utils.py
index 1b644cd7..908f8b86 100644
--- a/pylint/extensions/_check_docs_utils.py
+++ b/pylint/extensions/_check_docs_utils.py
@@ -161,12 +161,11 @@ def possible_exc_types(node: nodes.NodeNG) -> Set[nodes.ClassDef]:
                     continue
 
                 val = utils.safe_infer(ret.value)
-                if (
-                    val
-                    and isinstance(val, (astroid.Instance, nodes.ClassDef))
-                    and utils.inherit_from_std_ex(val)
-                ):
-                    excs.append(val)
+                if val and utils.inherit_from_std_ex(val):
+                    if isinstance(val, nodes.ClassDef):
+                        excs.append(val)
+                    elif isinstance(val, astroid.Instance):
+                        excs.append(val.getattr("__class__")[0])
 
     final_exceptions = set()
     for exc in excs:
@@ -176,9 +175,6 @@ def possible_exc_types(node: nodes.NodeNG) -> Set[nodes.ClassDef]:
                 continue
         except astroid.InferenceError:
             continue
-        if not isinstance(exc, nodes.ClassDef):
-            # needed for instances of classes
-            exc = exc.getattr("__class__")[0]
         final_exceptions.add(exc)
     return final_exceptions
 ```

Reply:
done